sign1.go

```diff
@@ -0,0 +1,230 @@
+//
+//   COSE_Sign1 = [
+//       Headers,
+//       payload : bstr / nil,
+//       signature : bstr
+//   ]
+//
+// Reference: https://tools.ietf.org/html/rfc8152#section-4.2
+type sign1Message struct {
+	_           struct{} `cbor:",toarray"`
+	Protected   cbor.RawMessage
+	Unprotected cbor.RawMessage
+	Payload     []byte
+	Signature   []byte
+}
+
+// sign1MessagePrefix represents the fixed prefix of COSE_Sign1_Tagged.
+var sign1MessagePrefix = []byte{
+	0xd2, // #6.18
+	0x84, // Array of length 4
+}
+
+// Sign1Message represents a decoded COSE_Sign1 message.
+//
+// Reference: https://tools.ietf.org/html/rfc8152#section-4.2
+type Sign1Message struct {
+	Headers   Headers
+	External  []byte
+	Payload   []byte
+	Signature []byte
+}
+
+// NewSign1Message returns a Sign1Message with header initialized.
+func NewSign1Message() *Sign1Message {
+	return &Sign1Message{
+		Headers: Headers{
+			Protected:   ProtectedHeader{},
+			Unprotected: UnprotectedHeader{},
+		},
+	}
+}
+
+// MarshalCBOR encodes Sign1Message into a COSE_Sign1_Tagged object.
+func (m *Sign1Message) MarshalCBOR() ([]byte, error) {
+	content := sign1Message{
+		Protected:   m.Headers.RawProtected,
+		Unprotected: m.Headers.RawUnprotected,
+		Payload:     m.Payload,
+		Signature:   m.Signature,
+	}
+	if content.Protected == nil {
+		header, err := encMode.Marshal(m.Headers.Protected)
+		if err != nil {
+			return nil, err
+		}
+		content.Protected = header
+	}
+	if content.Unprotected == nil {
+		header, err := encMode.Marshal(m.Headers.Unprotected)
+		if err != nil {
+			return nil, err
+		}
+		content.Unprotected = header
+	}
+	return encMode.Marshal(cbor.Tag{
+		Number:  CBORTagSign1Message,
+		Content: content,
+	})
+}
+
+// UnmarshalCBOR decodes a COSE_Sign1_Tagged object into Sign1Message.
+func (m *Sign1Message) UnmarshalCBOR(data []byte) error {
+	if m == nil {
+		return errors.New("cbor: UnmarshalCBOR on nil Sign1Message pointer")
+	}
+
+	// fast message check
+	if !bytes.HasPrefix(data, sign1MessagePrefix) {
+		return errors.New("cbor: invalid COSE_Sign1_Tagged object")
+	}
+
+	// decode to sign1Message and parse
+	var raw sign1Message
+	if err := decMode.Unmarshal(data[1:], &raw); err != nil {
+		return err
+	}
+	msg := Sign1Message{
+		Headers: Headers{
+			RawProtected:   raw.Protected,
+			RawUnprotected: raw.Unprotected,
+		},
+		Payload:   raw.Payload,
+		Signature: raw.Signature,
+	}
+	if err := decMode.Unmarshal(msg.Headers.RawProtected, &msg.Headers.Protected); err != nil {
+		return fmt.Errorf("cbor: invalid protected header: %w", err)
+	}
+	if err := decMode.Unmarshal(msg.Headers.RawUnprotected, &msg.Headers.Unprotected); err != nil {
+		return fmt.Errorf("cbor: invalid unprotected header: %w", err)
+	}
+
+	*m = msg
+	return nil
+}
+
+// Sign signs a Sign1Message using the provided Signer.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-4.4
+func (m *Sign1Message) Sign(rand io.Reader, signer Signer) error {
```

Comment:
@shizhMSFT have you thought about [this proposal](https://github.com/veraison/go-cose/pull/14#discussion_r836487473)?

IMO promoting the Sign and Verify method to a top-level function would make the API easier to use and have less footguns.

```go
func Sign1(rand io.Reader, signer Signer, payload, external []byte, headers Headers) (*Sign1Message, error)
func Verify1(msg *Sign1Message, verifier Verifier) error 
```

Reply:
I'm still thinking about what APIs look like for `COSE_Sign` since it should be consistent with `COSE_Sign1`.

Reply:
`COSE_Sign` has complex signing and verification scenarios. I will drop `Sign` and `Verify` but will adopt `Sign1` and `Verify1`.